src/data/real/sqrt.lean

```diff
@@ -0,0 +1,242 @@
+Authors: Mario Carneiro, Floris van Doorn, Yury Kudryashov
+-/
+import topology.instances.nnreal
+
+/-!
+# Square root of a real number
+
+In this file we define
+
+* `nnreal.sqrt` to be the square root of a nonnegative real number.
+* `real.sqrt` to be the square root of a real number, defined to be zero on negative numbers.
+
+Then we prove some basic properties of these functions.
+
+## Implementation notes
+
+We define `nnreal.sqrt` as the noncomputable inverse to the function `x ↦ x * x`. We use general
+theory of inverses of strictly monotone functions to prove that `nnreal.sqrt x` exists. As a side
+effect, `nnreal.sqrt` is a bundled `order_iso`, so for `nnreal` numbers we get continuity as well as
+theorems like `sqrt x ≤ y ↔ x * x ≤ y` for free.
+
+Then we define `real.sqrt x` to be `nnreal.sqrt (nnreal.of_real x)`. We also define a Cauchy
+sequence `real.sqrt_aux (f : cau_seq ℚ abs)` which converges to `sqrt (mk f)` but do not prove (yet)
+that this sequence actually converges to `sqrt (mk f)`.
+
+## Tags
+
+square root
+-/
+
+open set filter
+open_locale filter nnreal topological_space
+
+namespace nnreal
+
+variables {x y : ℝ≥0}
+
+/-- Square root of a nonnegative real number. -/
+@[pp_nodot] noncomputable def sqrt : ℝ≥0 ≃o ℝ≥0 :=
+order_iso.symm $ strict_mono.order_iso_of_surjective (λ x, x * x)
+  (λ x y h, mul_self_lt_mul_self x.2 h) $
+  surjective_of_continuous (continuous_id.mul continuous_id) tendsto_mul_self_at_top
+    (by simp [order_bot.at_bot_eq])
+
+lemma sqrt_eq_iff_sqr_eq : sqrt x = y ↔ y * y = x :=
+sqrt.to_equiv.apply_eq_iff_eq_symm_apply.trans eq_comm
+
+lemma sqrt_le_iff : sqrt x ≤ y ↔ x ≤ y * y :=
+sqrt.to_galois_connection _ _
+
+lemma le_sqrt_iff : x ≤ sqrt y ↔ x * x ≤ y :=
+(sqrt.symm.to_galois_connection _ _).symm
+
+@[simp] lemma sqrt_eq_zero : sqrt x = 0 ↔ x = 0 :=
+sqrt_eq_iff_sqr_eq.trans $ by rw [eq_comm, zero_mul]
+
+@[simp] lemma sqrt_zero : sqrt 0 = 0 := sqrt_eq_zero.2 rfl
+
+@[simp] lemma sqrt_one : sqrt 1 = 1 := sqrt_eq_iff_sqr_eq.2 $ mul_one 1
+
+@[simp] lemma mul_sqrt_self (x : ℝ≥0) : sqrt x * sqrt x = x :=
+sqrt.symm_apply_apply x
+
+@[simp] lemma sqrt_mul_self (x : ℝ≥0) : sqrt (x * x) = x := sqrt.apply_symm_apply x
+
+lemma sqrt_mul (x y : ℝ≥0) : sqrt (x * y) = sqrt x * sqrt y :=
+by rw [sqrt_eq_iff_sqr_eq, mul_mul_mul_comm, mul_sqrt_self, mul_sqrt_self]
+
+noncomputable def sqrt_hom : monoid_with_zero_hom ℝ≥0 ℝ≥0 := ⟨sqrt, sqrt_zero, sqrt_one, sqrt_mul⟩
+
+lemma sqrt_inv (x : ℝ≥0) : sqrt (x⁻¹) = (sqrt x)⁻¹ := sqrt_hom.map_inv' x
+
+lemma sqrt_div (x y : ℝ≥0) : sqrt (x / y) = sqrt x / sqrt y := sqrt_hom.map_div x y
+
+end nnreal
+
+namespace real
+
+/-- An auxiliary sequence of rational numbers that converges to `real.sqrt (mk f)`.
+Currently this sequence is not used in `mathlib`.  -/
+def sqrt_aux (f : cau_seq ℚ abs) : ℕ → ℚ
+| 0       := rat.mk_nat (f 0).num.to_nat.sqrt (f 0).denom.sqrt
+| (n + 1) := let s := sqrt_aux n in max 0 $ (s + f (n+1) / s) / 2
```

Comment:
Is there any reason not to delete this construction?

Reply:
This (together with an estimate on the rate of convergence) can be used to write a tactic that computes square roots (like the one we have for `pi`).

Reply:
I see.  Is `cau_seq` the right input type for such a tactic, though?  In practice, I imagine one would use the tactic on rational numbers.

Reply:
I left it as it was because I didn't want to think about the design. I guess, the one who will actually will write the tactic "approximate sqrt" will have to rewrite this.